torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1920,14 +1990,92 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
-                self.guards.append(ShapeGuard(expr, stack))
+                guard = ShapeGuard(expr, stack)
             elif concrete_val is sympy.false:
-                self.guards.append(ShapeGuard(sympy.Not(expr), stack))
+                guard = ShapeGuard(sympy.Not(expr), stack)
             else:
-                self.guards.append(
-                    ShapeGuard(sympy.Eq(expr, concrete_val), stack))  # type: ignore[arg-type]
+                guard = ShapeGuard(sympy.Eq(expr, concrete_val), stack)  # type: ignore[arg-type]
+
+            self.guards.append(guard)
+            self.refine_ranges(guard)
+
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
+        def simplify(expr: sympy.Expr) -> sympy.Expr:
+            """
+            Simplification specialized for range refinement.
+            """
+            return self.maybe_simplify_floordiv(self.simplify(expr))
+
+        def simplify_until(expr: sympy.Expr, max_iterations: int = 10) -> sympy.Expr:
+            """
+            Calls 'simplify' either until it does not change or until it reaches the
+            maximum number of iterations.
+            """
+            max_iterations = 10
+            for _ in range(max_iterations):
+                previous, expr = expr, simplify(expr)
+                if expr == previous:
+                    break
+            return expr
+
+        RELOP_MIRROR = {
+            sympy.Ge: sympy.Le,
+            sympy.Gt: sympy.Lt,
+            sympy.Le: sympy.Ge,
+            sympy.Lt: sympy.Gt,
+        }
+
+        # List of expressions to be processed.
+        # Here, we try considering both LHS and RHS by mirroring the
+        # original expression: a < b ==> b > a
+        exprs = [guard.expr]
+
+        if type(guard.expr) in RELOP_MIRROR:
+            exprs.append(RELOP_MIRROR[type(guard.expr)](guard.expr.rhs, guard.expr.lhs))  # type: ignore
+
+        for expr in exprs:
+            # First, try to simplify the left-hand side.
+            expr = simplify_until(expr)
+
+            # Filter the guards that:
+            #   1. are relational operations
+            #   2. have a symbol as the left-hand side
+            #   3. already have a range
+            if (
+                not isinstance(expr, sympy.Rel)
+                or not isinstance(expr.lhs, sympy.Symbol)
+                or expr.lhs not in self.var_to_range
+            ):
+                continue
+
+            # Update the value range of the left-hand side, if the
+            # right-hand side provides a better range.
+            symbol = expr.lhs
+
+            vr = self.var_to_range[symbol]
+            lower, upper = vr.lower, vr.upper
+
+            rhs_vr = sympy_interp(ValueRangeAnalysis, self.var_to_range, expr.rhs)  # type: ignore
+            lower_guard, upper_guard = self.var_to_guards.get(symbol, (None, None))
+
+            if lower < rhs_vr.lower and isinstance(expr, (sympy.Eq, sympy.Ge, sympy.Gt)):
```

Comment:
I suppose a comment here would make the bounds here clearer. Something like "Let's suppose that we have a preexisting range for x `[0, 100]`. Now, we issue a guard `x > y`, where the range for y is `[50, 150]`. Then, `lower = 0, rhs_vr.lower = 50` and therefore refinement can happen, refining x to `[50, 100]`, since x must be greater than y, but the lowest y could be is 50."